packages/build/src/plugins/child/mutations.js

```diff
@@ -84,6 +97,45 @@ const validateReadonlyProperty = function ({ method, keys, state: { topProxy },
   return Reflect[method](proxy, key, ...args)
 }
 
+// Retrieve normalized property name
+const getPropName = function (keys, key) {
+  return serializeKeys([...keys, key].map(normalizeDynamicProp))
+}
+
+// Some properties are user-defined, i.e. we need to replace them with a "*" token
+const normalizeDynamicProp = function (key, index, keys) {
+  return isArrayItem(key) || isDynamicObjectProp(keys, index) ? '*' : key
+}
+
+// Check it the value is a value item. In that case, we replace its indice by "*"
+const isArrayItem = function (key) {
+  return Number.isInteger(Number(key))
+}
+
+// Check if a property name is dynamic, such as `functions.{functionName}`
+const isDynamicObjectProp = function (keys, index) {
+  return (
+    index !== 0 &&
+    DYNAMIC_OBJECT_PROPS.has(serializeKeys(keys.slice(0, index))) &&
+    !NON_DYNAMIC_OBJECT_PROPS.has(serializeKeys(keys.slice(0, index + 1)))
+  )
+}
+
+// Properties with dynamic children
+const DYNAMIC_OBJECT_PROPS = new Set(['functions'])
+// Children properties which should not be considered dynamic
+const NON_DYNAMIC_OBJECT_PROPS = new Set([
```

Comment:
It's a bit unfortunate that this list of properties now lives in both `@netlify/config` (https://github.com/netlify/build/blob/main/packages/config/src/functions_config.js#L47-L53) and in `@netlify/build`. Do you think it would make sense for `@netlify/config` to export this information somehow so that we can centralise it, or would that be overkill?

Reply:
I don't think this is specific to that list actually. For example, any change of properties inside `@netlify/config` would require changes to this file too. 

One possible option would be to put this code inside `@netlify/config` but this would require plugin processes to load `@netlify/config`, which they have been able to avoid so far. Also, the logic is actually quite tied to build plugins, for example it imports the list of `EVENTS`.

I think we might want to try to just rename/change this file when `@netlify/config` publishes breaking changes, and re-visit if it becomes cumbersome. What do you think?

Reply:
I mentioned this because if we add another property to the functions configuration API, it will need to be added to that list in `@netlify/config` so that it stops being treated as a function name (i.e. a dynamic property) and starts being treated as a configuration property (i.e. static property). I believe the same change would need to be made in here, which is what I was hoping to avoid.

Having said that, we probably won't need to do that very often, so we should be fine as long as we make a note somewhere that these two places are interconnected.

In any case, this feels like something beyond the scope of this PR, so by all means feel free to merge it.